src/azure-cli-core/azure/cli/core/adal_authentication.py

```diff
@@ -77,17 +75,56 @@ def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
 
         _, token, full_token, _ = self._get_token(_try_scopes_to_resource(scopes))
 
-        try:
-            expires_on = full_token['expiresOn']
-            return AccessToken(token, int(datetime.datetime.strptime(expires_on, '%Y-%m-%d %H:%M:%S.%f').timestamp()))
-        except:  # pylint: disable=bare-except
-            pass  # To avoid crashes due to some unexpected token formats
-
-        try:
-            return AccessToken(token, int(full_token['expiresIn'] + time.time()))
-        except KeyError:  # needed to deal with differing unserialized MSI token payload
+        # NEVER use expiresIn (expires_in) as the token is cached and expiresIn will be already out-of date
+        # when being retrieved.
+
+        # User token entry sample:
+        # {
+        #     "tokenType": "Bearer",
+        #     "expiresOn": "2020-11-13 14:44:42.492318",
+        #     "resource": "https://management.core.windows.net/",
+        #     "userId": "test@azuresdkteam.onmicrosoft.com",
+        #     "accessToken": "eyJ0eXAiOiJKV...",
+        #     "refreshToken": "0.ATcAImuCVN...",
+        #     "_clientId": "04b07795-8ddb-461a-bbee-02f9e1bf7b46",
+        #     "_authority": "https://login.microsoftonline.com/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a",
+        #     "isMRRT": True,
+        #     "expiresIn": 3599
+        # }
+
+        # Service Principal token entry sample:
+        # {
+        #     "tokenType": "Bearer",
+        #     "expiresIn": 3599,
+        #     "expiresOn": "2020-11-12 13:50:47.114324",
+        #     "resource": "https://management.core.windows.net/",
+        #     "accessToken": "eyJ0eXAiOiJKV...",
+        #     "isMRRT": True,
+        #     "_clientId": "22800c35-46c2-4210-b8a7-d8c3ec3b526f",
+        #     "_authority": "https://login.microsoftonline.com/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a"
+        # }
+        if 'expiresOn' in full_token:
+            import datetime
+            expires_on_timestamp = int(_timestamp(
+                datetime.datetime.strptime(full_token['expiresOn'], '%Y-%m-%d %H:%M:%S.%f')))
```

Comment:
The default timezone here will be the local time zone. Who does set the expiresOn value ? CLI client or the service ? If it is the service, the time zone may be different and the timestamp may be inaccurate.

Reply:
AAD server sets `expires_in`:

https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#successful-response-1

```
{
    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik5HVEZ2ZEstZnl0aEV1Q...",
    "token_type": "Bearer",
    "expires_in": 3599,
    "scope": "https%3A%2F%2Fgraph.microsoft.com%2Fmail.read",
    "refresh_token": "AwABAAAAvPM1KaPlrEqdFSBzjqfTGAMxZGUTdM0t4B4...",
    "id_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJub25lIn0.eyJhdWQiOiIyZDRkMTFhMi1mODE0LTQ2YTctOD...",
}
```

and ADAL converts it to a local time:

https://github.com/AzureAD/azure-activedirectory-library-for-python/blob/35f9e003bed47936a1df3c6eb696691dc1fa91c3/adal/oauth2_client.py#L183-L187

```py
expires_in = wire_response.get(OAuth2.ResponseParameters.EXPIRES_IN)
if expires_in:
    now = datetime.now()
    soon = timedelta(seconds=expires_in)
    wire_response[OAuth2.ResponseParameters.EXPIRES_ON] = str(now + soon)
```

So `expiresOn` is computed locally and won't hit the "the time zone may be different and the timestamp may be inaccurate" issue.